framework/Models/EnsembleModel.py

```diff
@@ -221,61 +221,72 @@ def initialize(self,runInfo,inputs,initDict=None):
       @ Out, None
     """
-    # in here we store the job ids for which we did not collected the optional output yet
+    # store the job ids for jobs that we haven't collected optional output from
     self.tempOutputs['uncollectedJobIds'] = []
-    #self.tempOutputs['forHold'] = {}
     # collect name of all the outputs in the Step
     outputsNames = []
     if initDict is not None:
       outputsNames = [output.name for output in initDict['Output']]
 
     # here we check if all the inputs inputted in the Step containing the EnsembleModel are actually used
-    checkDictInputsUsage = {}
-    for inp in inputs:
-      checkDictInputsUsage[inp] = False
+    # -> but why?  Does this check prevent problems later?
+    checkDictInputsUsage = dict((inp,False) for inp in inputs)
 
-    for modelIn in self.assemblerDict['Model']:
-      self.modelsDictionary[modelIn[2]]['Instance'] = modelIn[3]
+    # collect the models
+    self.allOutputs = set()
+    for modelClass,modelType,modelName,modelInstance in self.assemblerDict['Model']:
+      self.modelsDictionary[modelName]['Instance'] = modelInstance
       inputInstancesForModel = []
-      for inputName in self.modelsDictionary[modelIn[2]]['Input']:
+      for inputName in self.modelsDictionary[modelName]['Input']:
         inputInstancesForModel.append(self.retrieveObjectFromAssemblerDict('Input',inputName))
         checkDictInputsUsage[inputInstancesForModel[-1]] = True
-      self.modelsDictionary[modelIn[2]]['InputObject'] = inputInstancesForModel
+      self.modelsDictionary[modelName]['InputObject'] = inputInstancesForModel
 
       # retrieve 'Output' objects, such as DataObjects, Databases
-      if self.modelsDictionary[modelIn[2]]['Output'] is not None:
+      if self.modelsDictionary[modelName]['Output'] is not None:
         outputInstancesForModel = []
-        for output in self.modelsDictionary[modelIn[2]]['Output']:
+        for output in self.modelsDictionary[modelName]['Output']:
           outputObject = self.retrieveObjectFromAssemblerDict('Output',output)
           if outputObject.name not in outputsNames:
-            self.raiseAnError(IOError, "The optional Output "+outputObject.name+" listed for Model "+modelIn[2]+" is not present among the Step outputs!!!")
+            self.raiseAnError(IOError, "The optional Output "+outputObject.name+" listed for Model "+modelName+" is not present among the Step outputs!!!")
           outputInstancesForModel.append(outputObject)
-        self.modelsDictionary[modelIn[2]]['OutputObject'] = outputInstancesForModel
+        self.modelsDictionary[modelName]['OutputObject'] = outputInstancesForModel
       else:
-        self.modelsDictionary[modelIn[2]]['OutputObject'] = []
-      self.modelsDictionary[modelIn[2]]['Instance'].initialize(runInfo,inputInstancesForModel,initDict)
+        self.modelsDictionary[modelName]['OutputObject'] = []
+
```

Comment:
can you remove this empty line?

Reply:
Hm.  I guess.  It doesn't seem to help the readability much to me, but I removed it.